internal/app/machined/pkg/runtime/v1alpha1/bootloader/sdboot/sdboot.go

```diff
@@ -120,29 +109,59 @@ func ProbeWithCallback(disk string, options options.ProbeOptions, callback func(
 				return err
 			}
 
+			if len(ukiFiles) == 0 {
+				return fmt.Errorf("no UKI files found in %q", filepath.Join(constants.EFIMountPoint, "EFI", "Linux"))
+			}
+
 			options.Log("sd-boot: found UKI files: %v", xslices.Map(ukiFiles, filepath.Base))
 
-			// here we handle a case when we boot of just kernel+initrd/uki/iso and we don't have a booted entry
-			// then we know we're booted of UKI
-			if bootedEntry == "" && len(ukiFiles) == 1 {
-				options.Log("sd-boot: no booted entry found, assuming the only UKI file: %s", ukiFiles[0])
+			// If we booted of UKI/Kernel+Initramfs/ISO Talos installer will always be run which
+			// sets the `LoaderEntryDefault` to the UKI file name, so either for reboot with Kexec or upgrade
+			// we will always have the UKI file name in the `LoaderEntryDefault`
+			// and we can use it to determine the default entry.
+			bootEntry, err := ReadVariable(LoaderEntryDefaultName)
+			if err != nil {
+				return err
+			}
+
+			options.Log("sd-boot: LoaderEntryDefault: %s", bootEntry)
+
+			if bootEntry == "" {
+				// If we booted of a Disk image, only `LoaderEntrySelected` will be set until we do an upgrade
+				// which will set the `LoaderEntryDefault` to the UKI file name.
+				// So for reboot with Kexec we will have to read the `LoaderEntrySelected`
+				// upgrades will always have `LoaderEntryDefault` set to the UKI file name.
+				loaderEntrySelected, err := ReadVariable(LoaderEntrySelectedName)
```

Comment:
I'm confused here, if I selected a non-default entry, should the selected take over?

Reply:
but see my point in the issue about which one we should actually overwrite

Reply:
I think we can do better during install, on install/upgrade we set `LoaderEntryOneShot` and if the next one boots to completion during the stage we write meta we can write the current value of `LoaderEntryOneShot` if kexec'ed or `LoaderEntrySelected` if a normal reboot to the `LoaderEntryDefault` value.